### PR TITLE
[commissioning] PSKd base32 encoding/decoding

### PR DESCRIPTION
**Pull request by GiedriusM**
Changes that implement base32-thread encoding and decoding for the joiner PSKd.
Should be considered work-in-progress until issues are discussed and resolved.

Issues:
* No official base32-format specification
* Test cases
* Multiple use and definitions of maximum PSK(d) length (OPENTHREAD_PSK_MAX_LENGTH, OT_PSKD_MAX_SIZE, OT_PSKC_MAX_SIZE, OT_COMMISSIONING_PASSPHRASE_MAX_SIZE, Dtls::kPskMaxLength)
* Should the same changes applied to commissioner credential?
* How to resolve otNodeApi (windows project) not having access to otBase32Decode/Encode. Should it be re-implemented in otApi?

### CODE REVIEW

**Comment by jwhui** [comment]
Thanks for working on this!

Some comments:

1. On the base32 spec: until a spec errata is processed, we should just roll forward with what we think is most sensible.  How about we allow a lengths for the Joining Device Credential between 6 and 32, inclusive.  If we do not end up with a multiple of 8 bits, simply pad the end with zeros.

2. Note that the Joiner Device Credential and Commissioner Credential are somewhat different. 
 The Commissioner Credential is a utf-8 passphrase with length 6 to 255, inclusive.  The Commissioner Credential is used to derive the PSKc using a key-stretching algorithm and its length is fixed at 16 bytes.

3. Given that the Thread Certification Test Harness passes the PSKd, I think we should try to maintain a path that allows the Thread Certification Test Harness to work unmodified (you may have noticed that the Travis CI checks are failing currently because of this).  This means we need to provide the ability to input both the Joiner Device Credential and the PSKd.  @librasungirl, do you have thoughts here?

4. The Commissioner Credential is really something used by an External Commissioner.  Thread devices all use the PSKc.  OpenThread already provides helper functions to derive the PSKc from the Commissioning Credential.

5. Regarding the Windows project, I think it's fine to keep the API passing the PSKd and PSKc in binary form.  We can add the base32 format at a later time.

**Comment by GiedriusM** [comment]
> 1. On the base32 spec: until a spec errata is processed, we should just roll forward with what we think is most sensible. How about we allow a lengths for the Joining Device Credential between 6 and 32, inclusive. If we do not end up with a multiple of 8 bits, simply pad the end with zeros.

Yes, the JDC length should be 6 to 32 characters as per Thread 1.1.1 spec. I was more concerned with duplicate or vague usage of the defined constants. Will take a better to look at each of them.
As for the format itself, it has to come in multiple of 40bits and it cannot simply be padded with zero-data as the padding is used to encode the length of the decoded data. Currently I see two possible solutions:
1) Use the current implementation, i.e. enforced 40bit alignment and '=' as a padding symbol.
2) Use implicit padding, i.e. treat the encoded string as if it was aligned to 40bit block with padding, but do not enforce this requirement (in decoding) or add any padding (in encoding). I guess that's what you meant, just in different words.

> 2. Note that the Joiner Device Credential and Commissioner Credential are somewhat different.
The Commissioner Credential is a utf-8 passphrase with length 6 to 255, inclusive. The Commissioner Credential is used to derive the PSKc using a key-stretching algorithm and its length is fixed at 16 bytes.

I'll leave the PSKc untouched in this PR.

> 3. Given that the Thread Certification Test Harness passes the PSKd, I think we should try to maintain a path that allows the Thread Certification Test Harness to work unmodified (you may have noticed that the Travis CI checks are failing currently because of this). This means we need to provide the ability to input both the Joiner Device Credential and the PSKd.

I'm not very familiar with the harness scripts, so correct me if I'm wrong. From a quick glance I see that it uses two interfaces node_api.py (for windows?) and node_cli.py (for linux cli?). The former uses calls to otNodeApi C++ functions, so it's simple (and logical) to pass the binary PSKd. The CLI api, however uses shell commands, which are not particularly friendly to random binary data. Current (and previous) code works only because the binary PSKd has all printable symbols, so I would assume that it's only logical that the CLI should use some form of ASCII encoding. This encoding could be provided in the interface scripts (i.e. changes only in node_cli.py functions `commissioner_add_joiner` and `joiner_start`), so the actual certification scripts would be unchanged.

> 5. Regarding the Windows project, I think it's fine to keep the API passing the PSKd and PSKc in binary form. We can add the base32 format at a later time.

otNodeApi project can be left with binary only PSKd support as mentioned in the previous paragraph. This should fix the appveyor build error. However I assume that changes in the libopenthread[_k] project should not be reverted.

**Comment by jwhui** [comment]
> 2) Use implicit padding, i.e. treat the encoded string as if it was aligned to 40bit block with padding, but do not enforce this requirement (in decoding) or add any padding (in encoding). I guess that's what you meant, just in different words.

Yes, the above is what I meant in my previous comment.

> This encoding could be provided in the interface scripts (i.e. changes only in node_cli.py functions `commissioner_add_joiner` and `joiner_start`), so the actual certification scripts would be unchanged.

For background, the [Thread Test Harness](http://graniteriverlabs.com/thread/) is maintained via the Thread Group by GRL.  OpenThread's interface to the Thread Test Harness is implemented in [tools/harness-thci/OpenThread.py](https://github.com/openthread/openthread/blob/master/tools/harness-thci/OpenThread.py).  There, you can see that the PSKd is indeed a string, but that string is interpreted as a raw binary string (i.e. ascii values).  I agree this is not ideal given that the PSKd should be an arbitrary binary value.  Until we can drive changes in the harness interface itself, we need to keep compatibility with it.

> otNodeApi project can be left with binary only PSKd support as mentioned in the previous paragraph. This should fix the appveyor build error. However I assume that changes in the libopenthread[_k] project should not be reverted.

Key for now is to fix the windows build.  If we need to expose the base32 support via otApi, we can do that later.

**Comment by GiedriusM** [comment]
As discussed, I've fixed the appveyor build by interpreting the given string as a binary PSK and changed to implicit base32 padding.
Currently only the travis test fails due to unacceptable PSK, but for this a decision need to be made how to handle CLI and binary PSK. I could do it as with otNoteApi and just use the string as the PSKd, but this would neglect the issue that it's impossible to pass true binary PSKd via the CLI and also the encode/decode would not be used anywhere except in the tests.

I've also took a look at different definitions that I mentioned before and some more issues arose.

First, there are three definitions regarding the PSK - `OPENTHREAD_PSK_MAX_LENGTH`, `OT_PSKD_MAX_SIZE` and `Dtls::kPskMaxLength`. My suggestion to use and interpret them as follows:
* `OPENTHREAD_PSK_MAX_LENGTH` - use as a joiner credential length, i.e. 32 symbols. This should probaly be renamed to be less vague, for example OT_JOINER_CREDENTIAL_MAX_SIZE and moved from windows into openthread headers.
* `OT_PSKD_MAX_SIZE` - use as a PSKd length, i.e. 20 bytes, this would replace kPskMaxLength in commisioner and joiner code.
* `Dtls::kPskMaxLength` - use as a generic DTLS PSK length, i.e. according to RFC4279 it should be 64 octets.

Second, NCP spinel and windows ioctl pass PSKd as a character string, so it would be good to have some feedback from someone who's responsible for those parts, namely in what format should the key be passed - as a string or a binary array and length. Seeming that this PR interferes with different sections of OpenThread I would not want it to break it.

Finally, somewhat unrelated to this PR, but more of a general concern is the 6 byte joiner credential minimum length. First thing is that 6 base32 characters do not correspond to whole number of binary bytes, i.e. 6*5bits/8bits = 3.75 bytes. Second, in case of a implicit-padding base32 encoding it's impossible to have 6 character string, i.e. 3 bytes are encoded in 5 characters and 4 bytes are encoded in 7, therefore skipping 6 characters altogether. Third, having 30bit PSK would constitute about 10^9 different keys and it is reasonable to expect of a non-govermental party capable of checking 10^5 passwords per second, which gives less than three hours of brute-force attack. Not to mention that a well-funded could to orders of magnitude faster attacks. I doubt this will have any effect as the Thread spec is out, but I would highly suggest increasing the minimum allowed joiner credential length.

**Comment by jwhui** [comment]
@GiedriusM, thanks for the updates.

1. Agree with changing `OPENTHREAD_PSK_MAX_LENGTH` to `OPENTHREAD_JOINER_CREDENTIAL_MAX_SIZE`.
2. Agree with setting `OPENTHREAD_PSKD_MAX_SIZE` to 20 bytes.
3. As for `Dtls::kPskMaxLength`, this should be the same as `OPENTHREAD_PSKD_MAX_SIZE`.  Note that RFC4279 does not apply here since we are using a new ciphersuite TLS_ECJPAKE_WITH_AES_128_CCM_8.
4. As for passing the PSKd around via APIs, it would be nice to have these as binary array.
5. Agree that the minimum Joiner Credential length does not correspond to a multiple of 8 bits, let's just conform as best as we can to the existing spec and pad it out.  I don't think we need to worry about the conversion from PSKd back to Joiner Credential.  What matters here is Joiner Credential to PSKd.
6. The primary reason for using the ECJ-PAKE cipher suite is to deal with low-entropy PSKs.  Note that the PSK is only used for establishing a session key.  In this case, the session key is a 128-bit AES key.

**Comment by GiedriusM** [comment]
Since it's unclear how and where the joiner credential should be used with openthread APIs and test environment, I've changed the APIs to use binary PSKd + length, removed all PSKd conversion (decoding) and added TODOs (with commented-out blocks of code) on the end-nodes of the API, where the PSK strings are passed in.

To summarize the changes:
* aPSKd string replaced with aPSKd byte array and length in the API calls
* Added/modified windows IOCTL structures to accommodate PSKd length.
* Added PSKd length checks in every API call (may be redundant in many cases).
* Corrected definitions/constants related to PSKd length
* Added base32-thread encode and decode implementation (only used in tests)
* Added foundation for base32 encode/decode unit tests (a standard with official test-cases is needed).

Also, the spinel protocol passes the PSKd as a string in **PROP_THREAD_JOINERS**. I would consider it a "core" layer, rather than UI, so I think it should send the binary PSKd with length, but was reluctant to do any changes as there's an ongoing standardization taking place. IMO, joiner credential decoding should only take place at the user interface level, so please consider changing **PROP_THREAD_JOINERS** data structure from **"A(t(ULE))"** to **"A(t(dLE))"**.

**Comment by GiedriusM** [comment]
> As for Dtls::kPskMaxLength, this should be the same as OPENTHREAD_PSKD_MAX_SIZE. Note that RFC4279 does not apply here since we are using a new ciphersuite TLS_ECJPAKE_WITH_AES_128_CCM_8.

What I should've mentioned that I had in mind, is that if the DTLS was to be reused, e.g. by the user coap application with end-to-end encryption, then the kPskMaxLength should comply with the RFC4279 and be at least 64 bytes.

**Comment by jwhui** [comment]
@GiedriusM, thanks again for continuing to drive this forward.

The latest change set looks good to me.  I think this is the right incremental change until we figure out the best path forward on integration with the Thread Test Harness.

Agree that Spinel should be updated to pass binary rather than a string.  Would you like to address that piece as well?

> What I should've mentioned that I had in mind, is that if the DTLS was to be reused, e.g. by the user coap application with end-to-end encryption, then the kPskMaxLength should comply with the RFC4279 and be at least 64 bytes.

Ah, your intent is now clear.  OpenThread does not currently expose APIs to support sharing of DTLS with the application (which I think would be a great enhancement).  I think making the proposed change above can be made along with this overall enhancement.  Does that work for you?

**Comment by GiedriusM** [comment]
> Agree that Spinel should be updated to pass binary rather than a string. Would you like to address that piece as well?

I've just started doing that, but realized, that this needs only a small change in openthread, but multiple changes in wpantund (third-party/spinel, NCPControl and DBUS layers). I think I'll do a PR to wpantund first and once both PRs are approved they could be merged at the same time.

> I think making the proposed change above can be made along with this overall enhancement.  Does that work for you?

Agreed, I'll just add a comment mentioning RFC.

**Comment by GiedriusM** [comment]
Related pull request for wpantund is https://github.com/openthread/wpantund/pull/242

File: src/ncp/ncp_base_ftd.cpp [hunk 1]
```diff
@@ -378,11 +379,12 @@ otError NcpBase::InsertPropertyHandler_THREAD_JOINERS(const uint8_t *aValuePtr,
...
-                           SPINEL_DATATYPE_UTF8_S     // PSK
-                           SPINEL_DATATYPE_UINT32_S   // Timeout
-                           SPINEL_DATATYPE_EUI64_S    // Extended address
+                           SPINEL_DATATYPE_DATA_WLEN_S  // PSK
+                           SPINEL_DATATYPE_UINT32_S     // Timeout
+                           SPINEL_DATATYPE_EUI64_S      // Extended address
                        ),
-                       &aPSKd,
+                       &pskd,
+                       &pskdLength,
```
Comment (abtink) [inline src/ncp/ncp_base_ftd.cpp]: I certainly agree that using encoding of  `DATA_WLEN` makes more sense here. I also appreciate that you made the corresponding changes/PR in `wpantund` and also updated the spinel documentation. 

This PR will change an existing Spinel property definition/format. Usually we only consider appending new elements/fields to end of an existing format (to keep it backward compatible) or we try to define a new property if the format needs to change completely (e.g., we ensured that the whitelist/blacklist spinel properties behave as before and added new capabilities through extensions/new properties when the OT MAC filter API got changed). 

That said, since this is relatively new definition, I think we can still change it. 

I think @pszkotak (and the Nordic team) contributed the original code for joiner/commissioner properties in 2fa7a13c1b59057ee30fd1f9bc8f358344df3dc9.  @pszkotak are you ok with this change? Thanks.

Comment (GiedriusM) [reply src/ncp/ncp_base_ftd.cpp]: I was hesitant to do the changes in the spinel (as can be seen in https://github.com/openthread/openthread/pull/2185#issuecomment-331646223), but jwhui seemed to agree with the proposed change. If there was a strong reason against, then it would not be a big problem to revert before the 6a67899 commit. In any case, I hope it's acceptable.

Comment (pszkotak) [reply src/ncp/ncp_base_ftd.cpp]: @abtink, changing the PROP_THREAD_JOINERS to "A(t(ULE))" to "A(t(dLE))" will break the backward compability, but if we make sure that wpantund gets aligned it shouldn't be a problem IMHO.
Passing the DATATYPE_DATA_WLEN seems neater than DATATYPE_UTF8 that we have now.

Could you confirm at which point would you like to perform the base32 conversion?
Is it going to take place before the wpantund sends it via spinel (less data would be sent, which seems like a good idea) or afterwards on the NCP itself?

Comment (abtink) [reply src/ncp/ncp_base_ftd.cpp]: Thanks @pszkotak . So looks like we all agree to allow the spinel property format change. Let's also merge the wpantund PR https://github.com/openthread/wpantund/pull/242 at same time...

**Comment by darconeous** [comment]
It's not clear to me what we are getting by converting the spinel type to be binary instead of keeping it as a string. Is the spec changing to not use strings?

File: src/ncp/ncp_base_ftd.cpp (hunk 1 above)
Comment (darconeous) [reply src/ncp/ncp_base_ftd.cpp]: I remain unconvinced, it has not been made clear to me (maybe I missed it) why this is a good idea.

File: doc/spinel-protocol-src/spinel-tech-thread.md [hunk 2]
```diff
@@ -249,12 +249,12 @@ The stable leader network data.
 ### PROP 5391: PROP_THREAD_JOINERS {#prop-thread-joiners}
 
 * Type: Insert/Remove Only (optionally Read-Write)
-* Packed-Encoding: `A(t(ULE))`
+* Packed-Encoding: `A(t(dLE))`
```
Comment (darconeous) [inline doc/spinel-protocol-src/spinel-tech-thread.md]: Please justify why this change is necessary or desirable.

Comment (GiedriusM) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: As per Thread spec, PSKd is a binary object, which can contain 0x00 bytes, i.e. null-characters. So far only ASCII strings were used as PSKds. This change is supposed to allow for non-ASCII PSKds.

Comment (jwhui) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: The Joining Device Credential is encoded using characters.  The PSKd is encoded in binary.

From the Thread 1.1.1 specification:

> Joining Device Credential: A human-scaled passphrase for use in authenticating that a new Joiner device is the correct one. ... This credential when specifically encoded in binary is referred to as the PSKd (Pre- Shared Key for the Device). A Joining Device Credential is encoded as uppercase alphanumeric characters (base32-thread: 0-9,A-Y excluding I,O,Q, and Z for readability) with a minimum length of 6 such characters and a maximum length of 32 such characters.

Comment (darconeous) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: Is the PSKd just a decoded Joining Device Credential?

Comment (jwhui) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: Yes.

Comment (darconeous) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: Then why not make this field the Joining Device Credential field?

Comment (darconeous) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: It seems like the obvious thing to do, especially if the base32 encoding hasn't even been finalized yet.

Comment (darconeous) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: Wouldn't it be easier to re-define the field to be the Joining Device Credential instead of the raw PSKd? I mean, the whole reason it was originally defined as a string in Spinel was because it was intended to be the "Joining Device Credential"

**Comment by jwhui** [comment]
This is being tracked in [SPEC-742](https://threadgroup.atlassian.net/browse/SPEC-742).  In my discussions with the Thread Group, there seems to be general agreement that the Joiner Device Credential and PSKd are the same in both information and encoding.  There does not seem to be any use case for the binary version.  As a result, I'll close this PR.
